logins-sql/src/engine.rs

```diff
@@ -25,6 +25,10 @@ pub struct PasswordEngine {
     db: LoginDb,
 }
 
+// TODO: determine if this is actually correct
```

Comment:
This scares me.  Are you confident these are correct?  I'm surprised that anything with a `LoginDb` inside it would be...

Reply:
This can go now, internally `call_with_result` uses `AssertUnwindSafe` on its callback.

To be completely clear, there should not be a way to violate memory safety because your code is not unwind-safe, this is at a different level of correctness (how messed up is our internal state after unwind).

That said, our `panic`-safety story is still somewhat in flux, and not ideal. It's not clear what to do about it to me. While 'assume everything is panic-safe' is clearly not great, I think it's what would happen anyway.

The first version of this lib required `std::panic::UnwindSafe` on the callback passed to `call_with_result`, and the end result is that you *need* to use `AssertUnwindSafe` on every use of an FFI binding that wraps a method that closes over a `&mut T` (note that this isn't true for `*mut T`, for whatever reason).

There's no way for you to actually say "I've thought about this code and yes it's mutable methods are unwind safe" (`RefUnwindSafe` and `UnwindSafe` do not apply to `&mut T`, only to `T` and `&T`), which means users of this lib would need to place `AssertUnwindSafe` on a large number of calls in practice, which IMO means there's a very low chance they'd actually consider its usage (I know I wouldn't).

Because of this, I lifted the `UnwindSafe` requirement for `call_with_result`/`call_with_output`.

There are other options:

- Abort on panic, which I disagree with fairly strongly.
- Poison on panic, which is probably good for some cases, but IMO not this one.
- Reconstruct internal state (e.g. reopen the DB) on panic. I'd like to do this in the long term, but our use of SQLCipher makes this more difficult -- we can't reopen the database since we probably don't have the key. Keeping the key and path around is of course possible, though.